### DISCUSSION TITLE
Pass an argument in the Language __call__() besides the text

### DISCUSSION THREAD

**Discussion by databill86** [explosion/spaCy#8194] [Help: Coding & Implementations] [answered]
Hi, 

Is there any way to pass a custom argument along with the text in the nlp(text) call ? 

I can't override the nlp call as the signature is as follows : 

```
class Language(object):

...

    def __call__(self, text, disable=[], component_cfg=None):
```

I need this (dynamic) argument in some of my custom spacy pipelines to process the text, and I need it inside my pipelines. 
I can't pass this argument in the pipeline configs because it's variable, and I can't use "as_tuples" because I will not have access to the argument inside my pipelines. 

Thanks.

**Comment by polm** [comment, marked as answer]
There is no standard way to do this and I've never heard of anyone doing it before, so you're kind of in uncharted territory here. 

One thing you could do is subclass Language and add a custom `__call__` implementation. I think you'd have to do something like this:

```
doc = self.make_doc(text) # same as original
doc._.stuff = stuff # this is your custom data
...
the rest of the original implementation
...
```

Another thing you could do, that may or may not be simpler, is to use a component to store data, and stick it early in the pipeline. Something like this:

```
# these are the same length
texts = ...
extras = ...

@Language.component("add_extra"):
def add_extra(doc):
    doc._.extra = extras.pop(0) # pop off the front of the list
    return doc
```

This is potentially brittle, but you could instead use a dict where the text is the key or something.

**Reply by databill86** [reply]
Thank you for you answer! the first solution seems good for me. 

Actually, it may be very useful to pass custom data on the fly to the nlp pipelines. There are a lot of cases when these data can't be passed in the pipeline config because it's dynamic:

```
nlp.add_pipe("custom_compo",
                    config={"language":            language,
                            "dynamic_arg":     dynamic_arg  # <- static
                            })

```
I have a lot of use cases like this.

**Reply by adrianeboyd** [reply]
As a note, some related issues are: #4160, #5698

To add to what Paul suggested, another workaround which makes it easier to work with the normal pipeline with `nlp.pipe` and batching is to override `Language.make_doc` like this:

```python
import spacy
from spacy.tokens import Doc
from spacy.language import Language


def custom_make_doc(self, info):
    Doc.set_extension("my_ext", default="", force=True)
    doc = self.tokenizer(info["text"])
    doc._.my_ext = info["my_ext"]
    return doc


Language.make_doc = custom_make_doc
nlp = spacy.blank("en")
doc = nlp({"text": "cats", "my_ext": "123"})
docs = nlp.pipe([{"text": "cats", "my_ext": "123"}])
```

I think this should work with both `nlp.__call__` and `nlp.pipe`, but with multiprocessing I think you might see some type errors if you check because `make_doc` doesn't have the expected signature.

Doing it this way would affect all pipelines in your script, so it might not be the right choice. You could only override `make_doc` for a custom language (or e.g. `English`) instead so that the parent class `Language` isn't modified.

Another option is to use a custom tokenizer, but the configuration and serialization would require a bit of effort. Here's what the tokenizer option could look like (for v2, without the v3 config setup and without serialization): https://github.com/explosion/spaCy/issues/5698#issuecomment-657747114

In the future, we do need better alternatives. Our current idea would be to allow `__call__` and `pipe` to take either `Doc` or `str` as the input and if the input is a `Doc`, then the tokenizer step is simply skipped. Alternate suggestions are welcome!
